packages/slate/src/models/node.js

```diff
@@ -198,11 +197,20 @@ class Node {
     first = assertKey(first)
     second = assertKey(second)
 
-    const keys = this.getKeysAsArray()
-    const firstIndex = keys.indexOf(first)
-    const secondIndex = keys.indexOf(second)
-    if (firstIndex == -1 || secondIndex == -1) return null
+    const firstAncestor = this.getFurthestAncestor(first)
+    const secondAncestor = this.getFurthestAncestor(second)
+    if (first === second) return false
+    if (!firstAncestor || !secondAncestor) {
+      return null
+    }
+    if (firstAncestor === secondAncestor) {
+      if (firstAncestor.key === first) return true
+      if (firstAncestor.key === second) return false
+      return firstAncestor.areDescendantsSorted(first, second)
+    }
 
+    const firstIndex = this.nodes.indexOf(firstAncestor)
+    const secondIndex = this.nodes.indexOf(secondAncestor)
```

Comment:
Maybe that's something we should come back to, after we worked through `getAncestors` and the likes. Maybe #1763 will do something about it ?

Reply:
I have put the `first === second` prior to the rest in the recent commit.  But there is something I am not sure:

Because slate memoize getAncestors, the recursive descendant way is almost as fast as the getCommonAncestor way.  However, if we use the recursive descendant method, all common ancestors will memoize the areSorted result.  

I think it perhaps depends on which one we want to cache, areSorted of all common ancestors., or getCommonAncestor

Reply:
Well, if `getCommonAncestor` is rewritten to use `getAncestors` (see [this PR](https://github.com/ianstormtaylor/slate/pull/1763/files#diff-54c7838b3ea52077be8fa46cee479d9bR596)), and we only cache `getAncestors`, then the choice is between caching `areSorted` on common ancestors, or caching `getAncestors` of both descendant. Caching `getAncestors` seems better, especially considering that the input space of `areSorted` is very large and less likely to yield cache hits (3 arguments to cache: `this`, `first`, `second`).

Reply:
OK I re-read your comment. `areSorted` does not seem worthy to memoize. I don't think we need to memoize `getCommonAncestor` either. Both methods have low complexity, since they scale with document depth, which is often shallow. Am I right ?

Reply:
My mistake.  I thought getFurthestAncestor calls getAncestors, which does not indeed.  I will use different ancestor version.

Reply:
I am confused about getCommonAncestors; why not use getAncestors in getCommonAncestors?